File: Anul II/Medii de proiectare si programare/Laborator/lab08/C#/Ubb.BikeContest.Networking/ProtocolBuffers/ProtocolBuilderUtils.cs
using Com.Ubb.Protocolbuffers;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Ubb.BikeContest.Model;

using protocol = Com.Ubb.Protocolbuffers;
using User = Ubb.BikeContest.Model.User;
using Participant = Ubb.BikeContest.Model.Participant;
using Race = Ubb.BikeContest.Model.Race;
using RaceEntry = Ubb.BikeContest.Model.RaceEntry;
using Team = Ubb.BikeContest.Model.Team;
using RaceDTO = Ubb.BikeContest.Model.data.RaceDto;
using System.Configuration;

namespace Ubb.BikeContest.Networking.ProtocolBuffers
{
    static class ProtocolBuilderUtils
    {
        public static BikeContestRequest CreateLoginRequest(string username, string passwordToken)
        {
            return new BikeContestRequest
            {
                Type = BikeContestRequest.Types.Type.Login,
                Username = username,
                PasswordToken = passwordToken
            };
        }

        public static BikeContestRequest CreateLogoutRequest(User user)
        {
            return new BikeContestRequest
            {
                Type = BikeContestRequest.Types.Type.Logout,
                User = BuildUser(user)
            };
        }

        public static BikeContestRequest CreateNewParticipantRequest(Participant participant)
        {
            return new BikeContestRequest
            {
                Type = BikeContestRequest.Types.Type.CreateParticipant,
                Participant = BuildParticipant(participant)
            };
        }

        public static BikeContestRequest CreateNewRaceEntriesRequest(List<RaceEntry> entries)
        {
            List<protocol.RaceEntry> buildEntries = new();
            entries.ForEach(entry => buildEntries.Add(BuildRaceEntry(entry)));

            return new BikeContestRequest
            {
                Type = BikeContestRequest.Types.Type.CreateRaceEntries,
                RaceEntries = { buildEntries }
            };
        }

        public static BikeContestRequest CreateNewRaceRequest(Race race)
        {
            return new BikeContestRequest
            {
                Type = BikeContestRequest.Types.Type.CreateRace,
                Race = BuildRace(race)
            };
        }

        public static BikeContestRequest CreateGetEntriesByRaceRequest(long raceId)
        {
            return new BikeContestRequest
            {
                Type = BikeContestRequest.Types.Type.GetEntriesByRace,
                RaceID = raceId
            };
        }

        public static BikeContestRequest CreateGetParticipantByDataRequest(Participant participant)
        {
            return new BikeContestRequest
            {
                Type = BikeContestRequest.Types.Type.GetParticipantByData,
                Participant = BuildParticipant(participant)
            };
        }

        public static BikeContestRequest CreateGetParticipantsByTeamRequest(long teamId)
        {
            return new BikeContestRequest
            {
                Type = BikeContestRequest.Types.Type.GetParticipantsByTeam,
                TeamID = teamId,
            };
        }

        public static BikeContestRequest CreateGetParticipantsRequest()
        {
            return new BikeContestRequest
            {
                Type = BikeContestRequest.Types.Type.GetParticipants
            };
        }

        public static BikeContestRequest CreateGetRaceByNameRequest(string raceName)
        {
            return new BikeContestRequest
            {
                Type = BikeContestRequest.Types.Type.GetRaceByName,
                RaceName = raceName,
            };
        }

        public static BikeContestRequest CreateGetRacesByEngineCapacityRequest(int capacity)
        {
            return new BikeContestRequest
            {
                Type = BikeContestRequest.Types.Type.GetRacesByEngineCapacity,
                EngineCapacity = capacity,
            };
        }

        public static BikeContestRequest CreateGetRacesWithParticipantCountRequest()
        {
            return new BikeContestRequest
            {
                Type = BikeContestRequest.Types.Type.GetRacesWithParticipantCount
            };
        }

        public static BikeContestRequest CreateGetTeamByNameRequest(string teamName)
        {
            return new BikeContestRequest
            {
                Type = BikeContestRequest.Types.Type.GetTeamByName,
                TeamName = teamName,
            };
        }

        public static BikeContestRequest CreateGetTeamsRequest()
        {
            return new BikeContestRequest
            {
                Type = BikeContestRequest.Types.Type.GetTeams
            };
        }

        public static BikeContestRequest CreateGetUnregisteredRacesRequest(
            long participantId, int engineCapacity)
        {
            return new BikeContestRequest
            {
                Type = BikeContestRequest.Types.Type.GetUnregisteredRaces,
                ParticipantID = participantId,
                EngineCapacity = engineCapacity
            };
        }

        public static BikeContestResponse CreateOkResponse()
        {
            return new BikeContestResponse
            {
                Type = BikeContestResponse.Types.Type.Ok
            };
        }

        public static BikeContestResponse CreateOkResponse(User user)
        {
            return new BikeContestResponse
            {
                Type = BikeContestResponse.Types.Type.Ok,
                User = BuildUser(user)
            };
        }

        public static BikeContestResponse CreateErrorResponse(string message)
        {
            return new BikeContestResponse
            {
                Type = BikeContestResponse.Types.Type.Error,
                ErrorMessage = message
            };
        }

        public static BikeContestResponse CreateAllParticipantsResponse(List<Participant> participants)
        {
            List<protocol.Participant> buildParticipants = new();
            participants.ForEach(participant => buildParticipants.Add(BuildParticipant(participant)));

            return new BikeContestResponse
            {
                Type = BikeContestResponse.Types.Type.AllParticipants,
                Participants = { buildParticipants }
            };
        }

        public static BikeContestResponse CreateAllTeamsResponse(List<Team> teams)
        {
            List<protocol.Team> buildTeams = new();
            teams.ForEach(team => buildTeams.Add(BuildTeam(team)));

            return new BikeContestResponse
            {
                Type = BikeContestResponse.Types.Type.AllTeams,
                Teams = { buildTeams }
            };
        }

        public static BikeContestResponse CreateEntriesByRaceResponse(List<RaceEntry> raceEntries)
        {
            List<protocol.RaceEntry> buildEntries = new();
            raceEntries.ForEach(entry => buildEntries.Add(BuildRaceEntry(entry)));

            return new BikeContestResponse
            {
                Type = BikeContestResponse.Types.Type.EntriesByRace,
                Entries = { buildEntries }
            };
        }

        public static BikeContestResponse CreateParticipantsByTeamResponse(List<Participant> participants)
        {
            List<protocol.Participant> buildParticipants = new();
            participants.ForEach(participant => buildParticipants.Add(BuildParticipant(participant)));

            return new BikeContestResponse
            {
                Type = BikeContestResponse.Types.Type.GetParticipantsByTeam,
                Participants = { buildParticipants }
            };
        }

        public static BikeContestResponse CreateNewParticipantResponse(Participant participant)
        {
            return new BikeContestResponse
            {
                Type = BikeContestResponse.Types.Type.NewParticipant,
                Participant = BuildParticipant(participant)
            };
        }

        public static BikeContestResponse CreateNewRaceResponse(RaceDTO race)
        {
            return new BikeContestResponse
            {
                Type = BikeContestResponse.Types.Type.NewParticipant,
                RaceDTO = BuildRaceDto(race)
            };
        }

        public static BikeContestResponse CreateParticipantResponse(Participant participant)
        {
            return new BikeContestResponse
            {
                Type = BikeContestResponse.Types.Type.GetParticipant,
                Participant = BuildParticipant(participant)
            };
        }

        public static BikeContestResponse CreateGetRaceByNameResponse(Race race)
        {
            return new BikeContestResponse
            {
                Type = BikeContestResponse.Types.Type.RaceByName,
                Race = BuildRace(race)
            };
        }

        public static BikeContestResponse CreateGetRacesByCapacityResponse(List<Race> races)
        {
            List<protocol.Race> buildRaces = new();
            races.ForEach(race => buildRaces.Add(BuildRace(race)));

            return new BikeContestResponse
            {
                Type = BikeContestResponse.Types.Type.RacesByCapacity,
                Races = { buildRaces }
            };
        }

        public static BikeContestResponse CreateGetRacesWithParticipantsResponse(List<RaceDTO> races)
        {
            List<protocol.RaceDTO> buildRaces = new();
            races.ForEach(race => buildRaces.Add(BuildRaceDto(race)));

            return new BikeContestResponse
            {
                Type = BikeContestResponse.Types.Type.RacesWithParticipants,
                RacesDTO = { buildRaces }
            };
        }

        public static BikeContestResponse CreateGetTeamResponse(Team team)
        {
            return new BikeContestResponse
            {
                Type = BikeContestResponse.Types.Type.GetTeam,
                Team = BuildTeam(team)
            };
        }

        public static BikeContestResponse CreateUpdatedRacesResponse(List<RaceDTO> races)
        {
            List<protocol.RaceDTO> buildRaces = new();
            races.ForEach(race => buildRaces.Add(BuildRaceDto(race)));

            return new BikeContestResponse
            {
                Type = BikeContestResponse.Types.Type.UpdatedRaces,
                RacesDTO = { buildRaces }
            };
        }

        public static string GetErrorMessage(BikeContestResponse response)
        {
            return response.ErrorMessage;
        }

        public static User GetUser(BikeContestRequest request)
        {
            protocol.User user = request.User;
            User result = new(user.Username, user.FullName, user.PasswordToken)
            {
                Id = user.Id
            };
            return result;
        }

        public static User GetUser(BikeContestResponse request)
        {
            protocol.User user = request.User;
            User result = new(user.Username, user.FullName, user.PasswordToken)
            {
                Id = user.Id
            };
            return result;
        }

        public static string GetUsername(BikeContestRequest request)
        {
            return request.Username;
        }

        public static string GetPasswordToken(BikeContestRequest request)
        {
            return request.PasswordToken;
        }

        public static long GetParticipantId(BikeContestRequest request)
        {
            return request.ParticipantID;
        }

        public static int GetEngineCapacity(BikeContestRequest request)
        {
            return request.EngineCapacity;
        }

        public static string GetTeamName(BikeContestRequest request)
        {
            return request.TeamName;
        }

        public static string GetRaceName(BikeContestRequest request)
        {
            return request.RaceName;
        }

        public static long GetTeamId(BikeContestRequest request)
        {
            return request.TeamID;
        }

        public static Participant GetParticipant(BikeContestRequest request)
        {
            return UnbuildParticipant(request.Participant);
        }

        public static Participant GetParticipant(BikeContestResponse response)
        {
            return UnbuildParticipant(response.Participant);
        }

        public static long GetRaceId(BikeContestRequest request)
        {
            return request.RaceID;
        }

        public static List<RaceEntry> GetRaceEntries(BikeContestRequest request)
        {
            List<RaceEntry> result = new();
            foreach (protocol.RaceEntry entry in request.RaceEntries)
            {
                result.Add(UnbuildRaceEntry(entry));
            }

            return result;
        }

        public static List<RaceEntry> GetRaceEntries(BikeContestResponse response)
        {
            List<RaceEntry> result = new();
            foreach (protocol.RaceEntry entry in response.Entries)
            {
                result.Add(UnbuildRaceEntry(entry));
            }

            return result;
        }

        public static Race GetRace(BikeContestRequest request)
        {
            return UnbuildRace(request.Race);
        }

        public static Race GetRace(BikeContestResponse response)
        {
            return UnbuildRace(response.Race);
        }

        public static RaceDTO GetRaceDto(BikeContestResponse response)
        {
            return UnbuildRaceDto(response.RaceDTO);
        }

        public static List<RaceDTO> GetRaceDtos(BikeContestResponse response)
        {
            List<RaceDTO> result = new();
            foreach (protocol.RaceDTO race in response.RacesDTO)
            {
                result.Add(UnbuildRaceDto(race));
            }

            return result;
        }

        public static List<Participant> GetParticipants(BikeContestResponse response)
        {
            List<Participant> result = new();
            foreach (protocol.Participant participant in response.Participants)
            {
                result.Add(UnbuildParticipant(participant));
            }

            return result;
        }

        public static List<Race> GetRaces(BikeContestResponse response)
        {
            List<Race> result = new();
            foreach (protocol.Race race in response.Races)
            {
                result.Add(UnbuildRace(race));
            }

            return result;
        }

        public static Team GetTeam(BikeContestResponse response)
        {
            return UnbuildTeam(response.Team);
        }

        public static List<Team> GetTeams(BikeContestResponse response)
        {
            List<Team> result = new();
            foreach (protocol.Team team  in response.Teams)
            {
                result.Add(UnbuildTeam(team));
            }

            return result;
        }

        private static protocol.Participant BuildParticipant(Participant participant)
        {
            return new protocol.Participant
            {
                Id = participant.Id,
                FirstName = participant.FirstName,
                LastName = participant.LastName,
                EngineCapacity = participant.EngineCapacity,
                TeamID = participant.Id
            };
        }

        private static protocol.User BuildUser(User user)
        {
            return new protocol.User
            {
                Id = user.Id,
                Username = user.Username,
                FullName = user.FullName,
                PasswordToken = user.Password
            };
        }

        private static protocol.Race BuildRace(Race race)
        {
            return new protocol.Race
            {
                Id = race.Id,
                Name = race.Name,
                EngineCapacity = race.EngineCapacity
            };
        }

        private static protocol.RaceEntry BuildRaceEntry(RaceEntry raceEntry)
        {
            return new protocol.RaceEntry
            {
                Id = raceEntry.Id,
                Participant = BuildParticipant(raceEntry.Participant),
                Race = BuildRace(raceEntry.Race)
            };
        }

        private static protocol.Team BuildTeam(Team team)
        {
            return new protocol.Team
            {
                Id = team.Id,
                Name = team.Name
            };
        }

        private static protocol.RaceDTO BuildRaceDto(RaceDTO raceDTO)
        {
            return new protocol.RaceDTO
            {
                Name = raceDTO.Name,
                EngineCapacity = raceDTO.EngineCapacity,
                Participants = raceDTO.Participants
            };
        }

        private static Participant UnbuildParticipant(protocol.Participant participant)
        {
            Participant result = new Participant(
                participant.FirstName, 
                participant.LastName, 
                participant.EngineCapacity);
            result.Id = participant.Id;
            result.TeamId = participant.TeamID;

            return result;
        }

        private static Race UnbuildRace(protocol.Race race)
        {
            Race result = new Race(
                race.Name, race.EngineCapacity);
            result.Id = race.Id;

            return result;
        }

        private static RaceDTO UnbuildRaceDto(protocol.RaceDTO raceDTO)
        {
            return new RaceDTO(
                raceDTO.Name, 
                raceDTO.EngineCapacity, 
                raceDTO.Participants);
        }

        private static RaceEntry UnbuildRaceEntry(protocol.RaceEntry raceEntry) 
        {
            RaceEntry result = new RaceEntry(
                UnbuildParticipant(raceEntry.Participant), 
                UnbuildRace(raceEntry.Race));
            result.Id = raceEntry.Id;

            return result;
        }

        private static Team UnbuildTeam(protocol.Team team)
        {
            Team result = new(team.Name)
            {
                Id = team.Id
            };

            return result;
        }
    }
}

File: Anul II/Medii de proiectare si programare/Laborator/lab08/Java/networking/src/main/java/com/ubb/protocolbuffers/ProtocolBuilderUtils.java
package com.ubb.protocolbuffers;

import com.ubb.dto.RaceDTO;
import com.ubb.model.*;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class ProtocolBuilderUtils {
    public static BikeContestProtocol.BikeContestRequest createLoginRequest(
            final String username, final String passwordToken) {

        return BikeContestProtocol.BikeContestRequest.newBuilder()
        .setType(BikeContestProtocol.BikeContestRequest.Type.Login)
        .setUsername(username)
        .setPasswordToken(passwordToken)
        .build();
    }

    public static BikeContestProtocol.BikeContestRequest createLogoutRequest(final User user) {
        return BikeContestProtocol.BikeContestRequest.newBuilder()
        .setType(BikeContestProtocol.BikeContestRequest.Type.Logout)
        .setUser(buildUser(user)).build();
    }

    public static BikeContestProtocol.BikeContestRequest createNewParticipantRequest(final Participant participant) {
        return BikeContestProtocol.BikeContestRequest.newBuilder()
        .setType(BikeContestProtocol.BikeContestRequest.Type.CreateParticipant)
        .setParticipant(buildParticipant(participant)).build();
    }

    public static BikeContestProtocol.BikeContestRequest createNewRaceEntriesRequest(final List<RaceEntry> entries) {
        final List<BikeContestProtocol.RaceEntry> buildEntries = new ArrayList<>();
        entries.forEach((entry) -> buildEntries.add(buildRaceEntry(entry)));

        return BikeContestProtocol.BikeContestRequest.newBuilder()
                .setType(BikeContestProtocol.BikeContestRequest.Type.CreateRaceEntries)
                .addAllRaceEntries(buildEntries).build();
    }

    public static BikeContestProtocol.BikeContestRequest createNewRaceRequest(final Race race) {
        return BikeContestProtocol.BikeContestRequest.newBuilder()
                .setType(BikeContestProtocol.BikeContestRequest.Type.CreateRace)
                .setRace(buildRace(race)).build();
    }

    public static BikeContestProtocol.BikeContestRequest createGetEntriesByRaceRequest(final Long raceId) {
        return BikeContestProtocol.BikeContestRequest.newBuilder()
                .setType(BikeContestProtocol.BikeContestRequest.Type.GetEntriesByRace)
                .setRaceID(raceId).build();
    }

    public static BikeContestProtocol.BikeContestRequest createGetParticipantByDataRequest(
            final Participant participant) {
        return BikeContestProtocol.BikeContestRequest.newBuilder()
                .setType(BikeContestProtocol.BikeContestRequest.Type.GetParticipantByData)
                .setParticipant(buildParticipant(participant)).build();
    }

    public static BikeContestProtocol.BikeContestRequest createGetParticipantsByTeamRequest(final Long teamId) {
        return BikeContestProtocol.BikeContestRequest.newBuilder()
                .setType(BikeContestProtocol.BikeContestRequest.Type.GetParticipantsByTeam)
                .setTeamID(teamId).build();
    }

    public static BikeContestProtocol.BikeContestRequest createGetParticipantsRequest() {
        return BikeContestProtocol.BikeContestRequest.newBuilder()
                .setType(BikeContestProtocol.BikeContestRequest.Type.GetParticipants).build();
    }

    public static BikeContestProtocol.BikeContestRequest createGetRaceByNameRequest(final String raceName) {
        return BikeContestProtocol.BikeContestRequest.newBuilder()
                .setType(BikeContestProtocol.BikeContestRequest.Type.GetRaceByName)
                .setRaceName(raceName).build();
    }

    public static BikeContestProtocol.BikeContestRequest createGetRacesByEngineCapacityRequest(
            final Integer capacity) {
        return BikeContestProtocol.BikeContestRequest.newBuilder()
                .setType(BikeContestProtocol.BikeContestRequest.Type.GetRacesByEngineCapacity)
                .setEngineCapacity(capacity).build();
    }

    public static BikeContestProtocol.BikeContestRequest createGetRacesWithParticipantCount() {
        return BikeContestProtocol.BikeContestRequest.newBuilder()
                .setType(BikeContestProtocol.BikeContestRequest.Type.GetRacesByEngineCapacity).build();
    }

    public static BikeContestProtocol.BikeContestRequest createGetTeamByNameRequest(final String teamName) {
        return BikeContestProtocol.BikeContestRequest.newBuilder()
                .setType(BikeContestProtocol.BikeContestRequest.Type.GetTeamByName)
                .setTeamName(teamName).build();
    }

    public static BikeContestProtocol.BikeContestRequest createGetTeamsRequest() {
        return BikeContestProtocol.BikeContestRequest.newBuilder()
                .setType(BikeContestProtocol.BikeContestRequest.Type.GetTeams).build();
    }

    public static BikeContestProtocol.BikeContestRequest createGetUnregisteredRacesRequest(
            final Long participantId, final Integer engineCapacity) {
        return BikeContestProtocol.BikeContestRequest.newBuilder()
                .setType(BikeContestProtocol.BikeContestRequest.Type.GetUnregisteredRaces)
                .setParticipantID(participantId)
                .setEngineCapacity(engineCapacity).build();
    }

    public static BikeContestProtocol.BikeContestResponse createOkResponse() {
        return BikeContestProtocol.BikeContestResponse.newBuilder()
        .setType(BikeContestProtocol.BikeContestResponse.Type.Ok).build();
    }

    public static BikeContestProtocol.BikeContestResponse createOkResponse(final User user) {
        return BikeContestProtocol.BikeContestResponse.newBuilder()
        .setType(BikeContestProtocol.BikeContestResponse.Type.Ok)
        .setUser(buildUser(user)).build();
    }

    public static BikeContestProtocol.BikeContestResponse createErrorResponse(final String message) {
        return BikeContestProtocol.BikeContestResponse.newBuilder()
                .setType(BikeContestProtocol.BikeContestResponse.Type.Error)
                .setErrorMessage(message).build();
    }

    public static BikeContestProtocol.BikeContestResponse createAllParticipantsResponse(
            final List<Participant> participants) {
        final List<BikeContestProtocol.Participant> buildParticipants = new ArrayList<>();
        participants.forEach((participant -> buildParticipants.add(buildParticipant(participant))));

        return BikeContestProtocol.BikeContestResponse.newBuilder()
                .setType(BikeContestProtocol.BikeContestResponse.Type.AllParticipants)
                .addAllParticipants(buildParticipants).build();
    }

    public static BikeContestProtocol.BikeContestResponse createAllTeamsResponse(final List<Team> teams) {
        final List<BikeContestProtocol.Team> buildTeams = new ArrayList<>();
        teams.forEach((team) -> buildTeams.add(buildTeam(team)));

        return BikeContestProtocol.BikeContestResponse.newBuilder()
                .setType(BikeContestProtocol.BikeContestResponse.Type.AllTeams)
                .addAllTeams(buildTeams).build();
    }

    public static BikeContestProtocol.BikeContestResponse createEntriesByRaceResponse(
            final List<RaceEntry> raceEntries) {
        final List<BikeContestProtocol.RaceEntry> buildEntries = new ArrayList<>();
        raceEntries.forEach((entry) -> buildEntries.add(buildRaceEntry(entry)));

        return BikeContestProtocol.BikeContestResponse.newBuilder()
                .setType(BikeContestProtocol.BikeContestResponse.Type.EntriesByRace)
                .addAllEntries(buildEntries).build();
    }

    public static BikeContestProtocol.BikeContestResponse createParticipantsByTeamResponse(
            final List<Participant> participants) {
        final List<BikeContestProtocol.Participant> buildParticipants = new ArrayList<>();
        participants.forEach((participant -> buildParticipants.add(buildParticipant(participant))));

        return BikeContestProtocol.BikeContestResponse.newBuilder()
                .setType(BikeContestProtocol.BikeContestResponse.Type.GetParticipantsByTeam)
                .addAllParticipants(buildParticipants).build();
    }

    public static BikeContestProtocol.BikeContestResponse createNewParticipantResponse(
            final Participant createdParticipant) {
        return BikeContestProtocol.BikeContestResponse.newBuilder()
                .setType(BikeContestProtocol.BikeContestResponse.Type.NewParticipant)
                .setParticipant(buildParticipant(createdParticipant)).build();
    }

    public static BikeContestProtocol.BikeContestResponse createNewRaceResponse(final RaceDTO race) {
        return BikeContestProtocol.BikeContestResponse.newBuilder()
                .setType(BikeContestProtocol.BikeContestResponse.Type.NewRace)
                .setRaceDTO(buildRaceDto(race)).build();
    }

    public static BikeContestProtocol.BikeContestResponse createParticipantResponse(final Participant participant) {
        return BikeContestProtocol.BikeContestResponse.newBuilder()
                .setType(BikeContestProtocol.BikeContestResponse.Type.GetParticipant)
                .setParticipant(buildParticipant(participant)).build();
    }

    public static BikeContestProtocol.BikeContestResponse createGetRaceByNameResponse(final Race race) {
        return BikeContestProtocol.BikeContestResponse.newBuilder()
                .setType(BikeContestProtocol.BikeContestResponse.Type.RaceByName)
                .setRace(buildRace(race)).build();
    }

    public static BikeContestProtocol.BikeContestResponse createGetRacesByCapacityResponse(final List<Race> races) {
        final List<BikeContestProtocol.Race> buildRaces = new ArrayList<>();
        races.forEach((race) -> buildRaces.add(buildRace(race)));

        return BikeContestProtocol.BikeContestResponse.newBuilder()
                .setType(BikeContestProtocol.BikeContestResponse.Type.RacesByCapacity)
                .addAllRaces(buildRaces).build();
    }

    public static BikeContestProtocol.BikeContestResponse createGetRacesWithParticipantsResponse(
            final List<RaceDTO> races) {
        final List<BikeContestProtocol.RaceDTO> buildRaces = new ArrayList<>();
        races.forEach((race) -> buildRaces.add(buildRaceDto(race)));

        return BikeContestProtocol.BikeContestResponse.newBuilder()
                .setType(BikeContestProtocol.BikeContestResponse.Type.RacesWithParticipants)
                .addAllRacesDTO(buildRaces).build();
    }

    public static BikeContestProtocol.BikeContestResponse createGetTeamResponse(final Team team) {
        return BikeContestProtocol.BikeContestResponse.newBuilder()
                .setType(BikeContestProtocol.BikeContestResponse.Type.GetTeam)
                .setTeam(buildTeam(team)).build();
    }

    public static BikeContestProtocol.BikeContestResponse createUpdatedRacesResponse(
            final List<RaceDTO> races) {
        final List<BikeContestProtocol.RaceDTO> buildRaces = new ArrayList<>();
        races.forEach((race) -> buildRaces.add(buildRaceDto(race)));

        return BikeContestProtocol.BikeContestResponse.newBuilder()
                .setType(BikeContestProtocol.BikeContestResponse.Type.UpdatedRaces)
                .addAllRacesDTO(buildRaces).build();
    }

    public static String getErrorMessage(final BikeContestProtocol.BikeContestResponse response) {
        return response.getErrorMessage();
    }

    public static User getUser(final BikeContestProtocol.BikeContestRequest request) {
        final BikeContestProtocol.User user = request.getUser();

        final User result = new User(user.getUsername(), user.getFullName());
        result.setId(user.getId());
        return result;
    }

    public static User getUser(final BikeContestProtocol.BikeContestResponse response) {
        final BikeContestProtocol.User user = response.getUser();

        final User result = new User(user.getUsername(), user.getFullName());
        result.setId(user.getId());
        return result;
    }

    public static String getUsername(final BikeContestProtocol.BikeContestRequest request) {
        return request.getUsername();
    }

    public static String getPasswordToken(final BikeContestProtocol.BikeContestRequest request) {
        return request.getPasswordToken();
    }

    public static Long getParticipantId(BikeContestProtocol.BikeContestRequest request) {
        return request.getParticipantID();
    }

    public static Integer getEngineCapacity(BikeContestProtocol.BikeContestRequest request) {
        return request.getEngineCapacity();
    }

    public static String getTeamName(BikeContestProtocol.BikeContestRequest request) {
        return request.getTeamName();
    }

    public static String getRaceName(BikeContestProtocol.BikeContestRequest request) {
        return request.getRaceName();
    }

    public static Long getTeamId(BikeContestProtocol.BikeContestRequest request) {
        return request.getTeamID();
    }

    public static Participant getParticipant(BikeContestProtocol.BikeContestRequest request) {
        return unbuildParticipant(request.getParticipant());
    }

    public static Participant getParticipant(BikeContestProtocol.BikeContestResponse response) {
        return unbuildParticipant(response.getParticipant());
    }

    public static Long getRaceId(BikeContestProtocol.BikeContestRequest request) {
        return request.getRaceID();
    }

    public static List<RaceEntry> getRaceEntries(BikeContestProtocol.BikeContestRequest request) {
        final List<RaceEntry> result = new ArrayList<>();
        request.getRaceEntriesList().forEach((buildEntry) -> result.add(unbuildRaceEntry(buildEntry)));

        return result;
    }

    public static List<RaceEntry> getRaceEntries(BikeContestProtocol.BikeContestResponse response) {
        final List<RaceEntry> result = new ArrayList<>();
        response.getEntriesList().forEach((buildEntry) -> result.add(unbuildRaceEntry(buildEntry)));

        return result;
    }

    public static Race getRaceDto(BikeContestProtocol.BikeContestRequest request) {
        return unbuildRace(request.getRace());
    }

    public static Race getRace(BikeContestProtocol.BikeContestResponse response) {
        return unbuildRace(response.getRace());
    }

    public static RaceDTO getRaceDto(BikeContestProtocol.BikeContestResponse response) {
        return unbuildRaceDto(response.getRaceDTO());
    }

    public static List<RaceDTO> getRaceDTOs(BikeContestProtocol.BikeContestResponse response) {
        final List<RaceDTO> result = new ArrayList<>();
        response.getRacesDTOList().forEach((raceDTO) -> result.add(unbuildRaceDto(raceDTO)));

        return result;
    }


    public static List<Participant> getParticipants(BikeContestProtocol.BikeContestResponse response) {
        List<Participant> result = new ArrayList<>();
        response.getParticipantsList().forEach((participant) -> {
            result.add(unbuildParticipant(participant));
        });

        return result;
    }

    public static List<Race> getRaces(BikeContestProtocol.BikeContestResponse response) {
        List<Race> result = new ArrayList<>();
        response.getRacesList().forEach((race) -> {
            result.add(unbuildRace(race));
        });

        return result;
    }

    public static Team getTeam(BikeContestProtocol.BikeContestResponse response) {
        return unbuildTeam(response.getTeam());
    }

    public static List<Team> getTeams(BikeContestProtocol.BikeContestResponse response) {
        List<Team> result = new ArrayList<>();
        response.getTeamsList().forEach((team) -> {
            result.add(unbuildTeam(team));
        });

        return result;
    }

    private static BikeContestProtocol.Participant buildParticipant(final Participant participant) {
        return BikeContestProtocol.Participant.newBuilder()
                .setId(participant.getId()).setFirstName(participant.getFirstName())
                .setLastName(participant.getLastName()).setEngineCapacity(participant.getEngineCapacity())
                .setTeamID(participant.getTeamID()).build();
    }

    private static BikeContestProtocol.User buildUser(final User user) {
        return BikeContestProtocol.User.newBuilder()
                .setId(user.getId()).setFullName(user.getFullName()).setUsername(user.getUsername())
                .setPasswordToken(user.getPasswordToken()).build();
    }

    private static BikeContestProtocol.Race buildRace(final Race race) {
        return BikeContestProtocol.Race.newBuilder()
                .setId(race.getId()).setName(race.getName()).setEngineCapacity(race.getEngineCapacity())
                .build();
    }

    private static BikeContestProtocol.RaceEntry buildRaceEntry(final RaceEntry raceEntry) {
        return BikeContestProtocol.RaceEntry.newBuilder()
                .setId(raceEntry.getId()).setParticipant(buildParticipant(raceEntry.getParticipant()))
                .setRace(buildRace(raceEntry.getRace())).build();
    }

    private static BikeContestProtocol.Team buildTeam(final Team team) {
        return BikeContestProtocol.Team.newBuilder()
                .setId(team.getId()).setName(team.getName()).build();
    }

    private static BikeContestProtocol.RaceDTO buildRaceDto(final RaceDTO raceDTO) {
        return BikeContestProtocol.RaceDTO.newBuilder()
                .setName(raceDTO.getName()).setEngineCapacity(raceDTO.getEngineCapacity())
                .setParticipants(raceDTO.getParticipants()).build();
    }

    private static Participant unbuildParticipant(final BikeContestProtocol.Participant participant) {
        final Participant result = new Participant(
                participant.getFirstName(),
                participant.getLastName(),
                participant.getEngineCapacity()
        );
        result.setId(participant.getId());

        return result;
    }

    private static Race unbuildRace(final BikeContestProtocol.Race race) {
        final Race result = new Race(
                race.getName(),
                race.getEngineCapacity()
        );
        result.setId(race.getId());

        return result;
    }

    private static RaceDTO unbuildRaceDto(final BikeContestProtocol.RaceDTO raceDTO) {
        return new RaceDTO(
                raceDTO.getName(),
                raceDTO.getEngineCapacity(),
                raceDTO.getParticipants()
        );
    }

    private static RaceEntry unbuildRaceEntry(final BikeContestProtocol.RaceEntry raceEntry) {
        final RaceEntry entry = new RaceEntry(
                unbuildParticipant(raceEntry.getParticipant()),
                unbuildRace(raceEntry.getRace())
        );
        entry.setId(raceEntry.getId());

        return entry;
    }

    private static Team unbuildTeam(final BikeContestProtocol.Team team) {
        final Team result = new Team(
                team.getName()
        );
        result.setId(team.getId());
        return result;
    }
}
